### ISSUE TITLE
[Bug]: How to navigate outside React context in v6?

### ISSUE THREAD

**Issue by hacker0limbo** [remix-run/react-router#8264]
### What version of React Router are you using?

v6

### Steps to Reproduce

In v6 docs, it mentions that we can use `useNavigate()` hook to do navigation, similar to in v5 we directly use `useHistory()` hook. However I am not sure how we can do the navigation outside React context in v6, cause in v5, your can manually provide a [`history`](https://v5.reactrouter.com/web/api/Router/history-object) object as a prop to `Router` component, and reference it in other places even it is not inside React context:

```js
// myHistory.js
import { createBrowserHistory } from "history";

const customHistory = createBrowserHistory();

export default  customHistory

// otherFile.js
import history from './myHistory.js'

function doSomething() {
  history.push('/xxx')
}
```

But how we could achieve the same functionality in v6? I have see a few posts asking similar questions on stackoverflow, but currently there is no solution provided:
- [React Router v6.0.0-alpha.5 history prop removal - navigating outside of react context](https://stackoverflow.com/q/61858941/12733140)
- [React Router ^6.0.0-beta.0 history prop removal - navigating outside of react context](https://stackoverflow.com/q/64192320/12733140)

### Expected Behavior

A common scenario from my experience was consider i have a redux thunk action creator that doing signup logic, and after sending request if success i wish the page can be navigate to home page:

```js
// signupActions.js

export const signupRequest = (userData) => {
  return dispatch => {
    dispatch(signupPending())

    return axios.post('/api/user', userData)
      .then(res => dispatch(signupSuccess()))
      .then(() => {
        // doing navigation
        // navigate('/')
      })
      .catch(error => dispatch(signupFailure(error.message)))
  }
}
```

The action is outside React context so i am not able to use `useNavigate()` hook, Although i can do some refactor and move some logic to the React component, but i prefer to keep most business logic inside action since i wish the component are more responsible for UI rendering.

### Actual Behavior

As mentioned above

**Comment by ryanflorence** [comment]
I'll make a quick guide on this, it's going to be a common question that we've already anticipated.

**Comment by ryanflorence** [comment]
Short answer: When your thunk is successful, change the state to something like `"success"` or `"redirect"` and then useEffect + navigate:

```tsx
export function AuthForm() {
  const auth = useAppSelector(selectAuth);
  const dispatch = useAppDispatch();
  const navigate = useNavigate();

  useEffect(() => {
    if (auth.status === "success") {
      navigate("/dashboard", { replace: true });
    }
  }, [auth.status, navigate]);

  return (
    <div>
      <button
        disabled={auth.status === "loading"}
        onClick={() => dispatch(login())}
      >
        {auth.status === "idle"
          ? "Sign in"
          : auth.status === "loading"
          ? "Signing in..."
          : null}
      </button>
    </div>
  );
}
```

**Comment by evenfrost** [comment]
@ryanflorence it's still doesn't answer the original question though, which is how to programmatically use navigation outside React components.

I believe it's not about one particular example that can be fixed with a different workflow, but a more general and pretty demanded functionality that is missing from the new version, if the library is to be called a 'fully-featured' one. Would love to see a guide on how it can be done.

Related: https://github.com/remix-run/react-router/pull/8284, https://github.com/remix-run/react-router/issues/7970.

**Comment by huczk** [comment]
I had this problem and created custom `HistoryRouter` like that:

```tsx
import { Update } from "history";
import { useLayoutEffect, useReducer } from "react";
import { Router } from "react-router-dom";

import { history } from "./history";

const reducer = (_: Update, action: Update) => action;

export const HistoryRouter: React.FC = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, {
    action: history.action,
    location: history.location,
  });

  useLayoutEffect(() => history.listen(dispatch), []);

  return (
    <Router navigationType={state.action} location={state.location} navigator={history}>
      {children}
    </Router>
  );
};
```

**Comment by amazecc** [comment]
source code reference： https://github1s.com/remix-run/react-router/blob/HEAD/packages/react-router-dom/index.tsx#L133

temporary plan, waiting for official support：

// history.ts
```typescript
import { createBrowserHistory } from "history";

export const history = createBrowserHistory();
```

// BrowserRouter.tsx
```typescript
import React from "react";
import { History } from "history";
import { BrowserRouterProps as NativeBrowserRouterProps, Router } from "react-router-dom";

export interface BrowserRouterProps extends Omit<NativeBrowserRouterProps, "window"> {
    history: History;
}

export const BrowserRouter: React.FC<BrowserRouterProps> = React.memo(props => {
    const { history, ...restProps } = props;
    const [state, setState] = React.useState({
        action: history.action,
        location: history.location,
    });

    React.useLayoutEffect(() => history.listen(setState), [history]);

    return <Router {...restProps} location={state.location} navigationType={state.action} navigator={history} />;
});

```

**Comment by predragnikolic** [comment]
@amazecc thanks for your comment :)

It works 👍 

<details>
<summary>
history.ts
</summary>

```ts
import { createBrowserHistory } from "history"

export const myHistory = createBrowserHistory({ window })
```
</details>

<details>
<summary>
HistoryRouter.ts
</summary>

```ts
// implementation from https://github1s.com/remix-run/react-router/blob/HEAD/packages/react-router-dom/index.tsx#L133-L137
import { ReactNode } from "react"
import { useLayoutEffect, useState } from "react"
import { History } from "history"
import { Router } from "react-router-dom"

export interface BrowserRouterProps {
  basename?: string;
  children?: ReactNode;
  history: History;
}
export function HistoryRouter({
  basename,
  children,
  history
}: BrowserRouterProps) {
  let [state, setState] = useState({
    action: history.action,
    location: history.location
  })

  useLayoutEffect(() => history.listen(setState), [history])

  return (
    <Router
      basename={basename}
      children={children}
      location={state.location}
      navigationType={state.action}
      navigator={history} />
  )
}
```
</details>

Than all you need is to wrap the `<App />` with `<HistoryRouter/>`.


```ts
import { HistoryRouter } from "./HistoryRouter"
import { myHistory } from "./history"

ReactDOM.render(
    <HistoryRouter history={myHistory}>
        <App />
    </HistoryRouter>
  document.getElementById("root")
)
```

and use your `myHistory` where you want:
```ts
import axios from "axios"
import { myHistory } from "@/router/history"

// Configure axios instance
const backendApi = axios.create(...)

backendApi.interceptors.response.use(function(response) {
  return response
}, async function (error) {

  if (error.response?.status === 403) {
    myHistory.replace(`/forbidden`) // Usage example.
    return Promise.reject(error) 
  }

  return Promise.reject(error)
})
```

**Comment by ayalpani** [comment]
> @amazecc and @huczk thanks for your comments :)
> 
> It works +1

Great, thanks from me too. I am actually using myHistory.replace/.push INSIDE of react components, as it does not trigger a re-render of all components (like useNavigate does), when the route changes.

**Comment by nimo-juice** [comment]
@ryanflorence do you think this is something react router v6 will support natively? We are also running into this issue. Crucial use case for us: if an API returns 401, token has expired, we want to redirect user to login screen with a message. We previously were able to do this pretty easily. This has become challenging now that the history object is no longer being exposed (and we can't use hooks in the API request code block)

**Comment by hsbtr** [comment]
I also encountered this problem, the official seems to be reluctant to support the use of navigation outside the context of react

**Comment by c0d3x** [comment]
What is the official answer for dealing with this issue?

**Comment by timdorr** [comment]
You can now use HistoryRouter (as of version 6.1.0) to maintain a global history instance that you can access anywhere:

```js
import { createBrowserHistory } from 'history';
import { HistoryRouter } from 'react-router-dom';

let history = createBrowserHistory();

function App() {
  return (
    <HistoryRouter history={history}>
      // The rest of your app
    </HistoryRouter>
  );
}

history.push("/foo");
```

**Comment by c0d3x** [comment]
> ```js
> createBrowserHistory
> ```

Does this also support hashRouter which I am using?

**Comment by mkarajohn** [comment]
@timdorr According to the docs, the `history` package is no longer a peer dependency we need to have installed and it's rather a direct dependency of `react-router-dom` [link](https://reactrouter.com/docs/en/v6/upgrading/v5#:~:text=You%27ll%20also%20want%20to%20remove%20the%20history%20dependency%20from%20your%20package) 

So you are suggesting that we import from a transitive dependency, which I do not like very much as a solution. (pnpm would also throw an error on this)

What if the `react-router-dom` package exported the `history` object itself instead?

**Comment by Gil-Porsansky** [comment]
@timdorr 
Hey can you please elaborate more on  unstable_HistoryRouter, is this the official solution from react router?
why is it called unstable_HistoryRouter, why unstable?, we need to know if we need to downgrade to v5 or to stay at 6.

**Comment by ruettenm** [comment]
Today I had to migrate react-router 5 to 6 and had the same question.

Before we used `<Router history={} />` because we also use the history in a mobx store outside.

Now I'm asking myself if I't the way to go to use `unstable_HistoryRouter` for this case or if this is bad practice?! I couldn't find more information what exactly "unstable" means :)

**Comment by toxxiczny** [comment]
what about this solution https://stackoverflow.com/a/70002872/10277045?

**Comment by timdorr** [comment]
The `unstable_` prefix just signifies you could encounter bugs due to mismatched versions of `history` from what `react-router` requests itself. While the practical implications are probably going to just create type issues (different types between versions of `history`), there could also be API or behavior changes that cause more subtle bugs (such as how URLs are parsed, or how paths are generated). 

As long as you're keeping an eye on the version of `history` requested by `react-router` and ensuring that's in sync with the version used by your app, you should be free of issues. We don't yet have protections or warnings against the mismatch, hence we're calling it unstable for now.

**Comment by davesteinberg** [comment]
Many thanks for adding `unstable_HistoryRouter`! I have projects with a lot of tests that depend on checking history to determine that an isolated component has done the right thing, and I'm excited that I should be able to bring them over to React Router 6 now.

I've run into a roadblock, though, wherever I have a component that uses `Navigate`. If I try to render such a component inside a `HistoryRouter`, with no `Routes`/`Route` in between, the render hangs.

I've created a minimal test case and opened bug #8591.

**Comment by shamseerahammedm** [comment]
> You can now use HistoryRouter (as of version 6.1.0) to maintain a global history instance that you can access anywhere:
> 
> ```js
> import { createBrowserHistory } from 'history';
> import { unstable_HistoryRouter as HistoryRouter } from 'react-router-dom';
> 
> let history = createBrowserHistory();
> 
> function App() {
>   return (
>     <HistoryRouter history={history}>
>       // The rest of your app
>     </HistoryRouter>
>   );
> }
> 
> history.push("/foo");
> ```

**React Router v6 introduces a new navigation API that is synonymous with <Link> and provides better compatibility with suspense-enabled apps.** 

as described in the official docs, does this mean that i'd still be using history api and not navigate api, so there is now way to do it with navigate api outside react components ?

**Comment by nmoinvaz** [comment]
Currently, I am using a custom history object to preserve query string between navigations as mentioned in this StackOverflow [article](https://stackoverflow.com/questions/43613140). Is using a custom history object with `HistoryRouter` the recommended way to preserve query string or is there another way?

**Comment by allicanseenow** [comment]
This is a really common use case, so do we have a solution for this? Ideally, something that could be done in 1 or 2 lines of code like how it should be like, instead of requiring devs to moving backwards to `unstable_HistoryRouter`, which is obviously something v6 discourages devs from doing?

**Comment by allicanseenow** [comment]
Can we reopen this issue?

**Comment by hexsprite** [comment]
This is what I came up with for a simple solution... seems to work OK.

```
export let extNavigate: (
  value: string | PromiseLike<string>
) => void | undefined
const useExtNavigate = () => {
  const navigate = useNavigate()

  useEffect(() => {
    ;(async () => {
      while (true) {
        const navigatePath = await new Promise<string>((resolve) => {
          extNavigate = resolve
        })
        navigate(navigatePath)
      }
    })()
  }, [navigate])
}
```

**Comment by oscarm3l1n** [comment]
> This is what I came up with for a simple solution... seems to work OK.
> 
> ```tsx
> export let extNavigate: (
>   value: string | PromiseLike<string>
> ) => void | undefined
> const useExtNavigate = () => {
>   const navigate = useNavigate()
> 
>   useEffect(() => {
>     ;(async () => {
>       while (true) {
>         const navigatePath = await new Promise<string>((resolve) => {
>           extNavigate = resolve
>         })
>         navigate(navigatePath)
>       }
>     })()
>   }, [navigate])
> }
> ```
> 
> Use by calling `extNavigate('/path')`

I'm getting the following error using your code: TypeError: extNavigate is not a function
    at Object.login

**Comment by GeorgiyBeloklokov** [comment]
> This is what I came up with for a simple solution... seems to work OK.
> 
> ```tsx
> export let extNavigate: (
>   value: string | PromiseLike<string>
> ) => void | undefined
> const useExtNavigate = () => {
>   const navigate = useNavigate()
> 
>   useEffect(() => {
>     ;(async () => {
>       while (true) {
>         const navigatePath = await new Promise<string>((resolve) => {
>           extNavigate = resolve
>         })
>         navigate(navigatePath)
>       }
>     })()
>   }, [navigate])
> }
> ```
> 
> Use by calling `extNavigate('/path')`

![Screenshot_11](https://user-images.githubusercontent.com/77876368/170288540-7407f7f5-f5de-4030-90f2-bccdd0c9ed7a.png)

**Comment by dongyuwei** [comment]
for `"react-router": "^6.2.2"`, you can use `Navigate()` method, eg:
```
import { Navigate } from 'react-router-dom';

axios.interceptors.response.use(
    (config) => {
        return config;
    },
    (error) => {
        if (error.response.data.statusCode === 401) {
            Navigate({ to: '/login' });
        }
        return Promise.reject(error);
    }
);
```

**Comment by allicanseenow** [comment]
Navigate is a component, not a function. I don't think you can use it like that.
https://reactrouter.com/docs/en/v6/components/navigate

**Comment by dongyuwei** [comment]
> Navigate is a component, not a function. I don't think you can use it like that. https://reactrouter.com/docs/en/v6/components/navigate

Thanks for your reply, I just deleted my comment and double checked my code, and find out the real workaround is the following :

```ts
// api.ts
axios.interceptors.response.use(
    (config) => {
        return config;
    },
    (error) => {
        if (error.response.data.statusCode === 401) {
            localStorage.removeItem('userInfo');
        }
        return Promise.reject(error);
    }
);
```

then use `useEffect` in a global functional component, eg HomeView to detect if the app need to redirect to '/login' page ,

```
export default function HomeView() {
    const navigate = useNavigate();

    useEffect(() => {
        const userInfo = localStorage.getItem('userInfo');
        if (!userInfo) {
            navigate('/login');
        }
    }, []);

    return <LayoutView />;
}
```

Let me summarize the key idea: even if you can't do navigation outside React context use react-router v6, but you can change a shared state(in our case, it's localStorage.userInfo) there, then you can do navigation in a global/shared functional component.

**Comment by luwanhuang** [comment]
> > Navigate is a component, not a function. I don't think you can use it like that. https://reactrouter.com/docs/en/v6/components/navigate
> 
> Thanks for your reply, I just deleted my comment and double checked my code, and find out the real workaround is the following :
> 
> ```ts
> // api.ts
> axios.interceptors.response.use(
>     (config) => {
>         return config;
>     },
>     (error) => {
>         if (error.response.data.statusCode === 401) {
>             localStorage.removeItem('userInfo');
>         }
>         return Promise.reject(error);
>     }
> );
> ```
> 
> then use `useEffect` in a global functional component, eg HomeView to detect if the app need to redirect to '/login' page ,
> 
> ```
> export default function HomeView() {
>     const navigate = useNavigate();
> 
>     useEffect(() => {
>         const userInfo = localStorage.getItem('userInfo');
>         if (!userInfo) {
>             navigate('/login');
>         }
>     }, []);
> 
>     return <LayoutView />;
> }
> ```
> 
> Let me summarize the key idea: even if you can't do navigation outside React context use react-router v6, but you can change a shared state(in our case, it's localStorage.userInfo) there, then you can do navigation in a global/shared functional component.

Hi, are you sure this code working? remove 'useInfo' shouldn't trigger that useEffect

**Comment by dongyuwei** [comment]
@luwanhuang The above code only works when user switch to new router path.

We have refactored our code, use  `window.postMessage('NotAuthorized');` in axios.interceptors code, then  listen to the message and do navigation. 
```
export default function HomeView() {
    const navigate = useNavigate();
    const location = useLocation();

    function onNotAuthorized(event: MessageEvent) {
        if (event.data === NotAuthorized) {
            navigate('/login');
        }
    }

    useEffect(() => {
        sessionStorage.setItem('lastPage', location.pathname);

        window.addEventListener('message', onNotAuthorized, false);

        return () => {
            window.removeEventListener('message', onNotAuthorized, false);
        };
    }, [location]);

    return <LayoutView />;
}
```

**Comment by aminify** [comment]
@timdorr after this much time has passed, I think it's not really good to keep the `unstable_` flag for the `HistoryRouter`. Why not just re-exporting `history` or `createBrowserHistory` in `react-router-dom`? This way devs can be sure they are using the same `history` version as react-router because it is coming directly from there.

**Comment by d-3-n** [comment]
> Navigate is a component, not a function. I don't think you can use it like that. https://reactrouter.com/docs/en/v6/components/navigate

useNavigate() hook returns a function. One can also use it as a component. Both ways are demonstrated in the upgrade guide: [https://reactrouter.com/en/v6.3.0/upgrading/v5]
Btw it seems components docu for RRv6 has been taken offline? All links return 404.

**Comment by Shaker-Pelcro** [comment]
What is the last update on this? How to redirect from a redux slice?

**Comment by andrey779evseev** [comment]
Before react router v6.4, I used this way to navigating outside of components.
```
import {createBrowserHistory} from 'history'

const history = createBrowserHistory({ window });

export const rootNavigate = (to: string) => {
  history.push(to);
};

createRoot(document.getElementById('root') as HTMLElement).render(
  <React.StrictMode>
      <HistoryRouter history={history}>
          <App />
      </HistoryRouter>
  </React.StrictMode>
);
```

But after update, history router no more available.

Is there a way now to navigating outside of components, without rollback to 6.3 version?

**Comment by denggj28** [comment]
have any offical sulution?
history module is not available in v 6.4

**Comment by Maadtin** [comment]
> Before react router v6.4, I used this way to navigating outside of components.
> 
> ```
> import {createBrowserHistory} from 'history'
> 
> const history = createBrowserHistory({ window });
> 
> export const rootNavigate = (to: string) => {
>   history.push(to);
> };
> 
> createRoot(document.getElementById('root') as HTMLElement).render(
>   <React.StrictMode>
>       <HistoryRouter history={history}>
>           <App />
>       </HistoryRouter>
>   </React.StrictMode>
> );
> ```
> 
> But after update, history router no more available.
> 
> Is there a way now to navigating outside of components, without rollback to 6.3 version?



> have any offical sulution? history module is not available in v 6.4

I think you can use <Router /> component with `navigator` prop.

**Comment by Oikio** [comment]
As I could not find a satisfying way to make it work, decision was to create a hook which will be connected to a global routing state and act like a proxy. 

Maybe not the best way to achieve it, but works well in our project. One of the benefits of such method - it's unobtrusive, you can keep React ecosystem separately from business logic.

```
type RoutingState = {
  currentUrl?: string
  desired?: {
    /** Ignores other fields if provided */
    goBack?: boolean
    /** Replace instead of push */
    replace?: boolean
    url: string
  }
}
```

```
export const useRoutingForUseCases = () => {
  const state = yourRoutingState()
  const navigate = useNavigate()
  const routerLocation = useLocation()

  useEffect(() => {
    if (!state) return

    if (state.desired?.goBack) {
      navigate(-1)
      return
    }

    if (state.desired && state.currentUrl !== state.desired.url) {
      navigate(state.desired.url, { replace: state.desired.replace })
    }
  }, [state?.desired])

  useEffect(() => {
    setYourRoutingState({
      url: `${routerLocation.pathname}${routerLocation.search}`
    })
  }, [routerLocation.pathname, routerLocation.search])
}
```

**Comment by sergioavazquez** [comment]
Is React Router team planning to provide an alternative for the missing `createBrowserHistory` method in v6.4? 
Being able to control navigation outside React components is something you just need in some cases. Hooks are great, but we can't use them everywhere. Do I need to find a new router for my projects??

> Before react router v6.4, I used this way to navigating outside of components.
> 
> ```
> import {createBrowserHistory} from 'history'
> 
> const history = createBrowserHistory({ window });
> 
> export const rootNavigate = (to: string) => {
>   history.push(to);
> };
> 
> createRoot(document.getElementById('root') as HTMLElement).render(
>   <React.StrictMode>
>       <HistoryRouter history={history}>
>           <App />
>       </HistoryRouter>
>   </React.StrictMode>
> );
> ```
> 
> But after update, history router no more available.
> 
> Is there a way now to navigating outside of components, without rollback to 6.3 version?

**Comment by evoyy** [comment]
@sergioavazquez Controlling navigation from outside the router is still possible, you just need to maintain a separate `history` dependency yourself (it's not a big problem). See: https://github.com/remix-run/react-router/issues/8264#issuecomment-991271554

**Comment by andrey779evseev** [comment]
> @sergioavazquez Controlling navigation from outside the router is still possible, you just need to maintain a separate `history` dependency yourself (it's not a big problem). See: [#8264 (comment)](https://github.com/remix-run/react-router/issues/8264#issuecomment-991271554)

Ok, but how to link history from this dependency to Router, without “unstable_HistoryRouter”? Or it’s no more needed? It will work without it?

**Comment by Arsikod** [comment]
Hi Remix team, any official solution on navigating from outside of component?

**Comment by evoyy** [comment]
@andrey779evseev @Arsikod The oofifical solution is the one described in https://github.com/remix-run/react-router/issues/8264#issuecomment-991271554 using `unstable_HistoryRouter`, and my understanding is that `unstable_HistoryRouter` is NOT going to be removed, it is prefixed with `unstable_` only to indicate that there is a possibility of your external history dependency losing compatibility with the histroy dependency used internally by react-router, so you must ensure that the versions match when you update react-router. You can find out which version of history you need by looking at react router's package.json https://github.com/remix-run/react-router/blob/main/package.json

**Comment by 9jaGuy** [comment]
What's the migration path for people doing this already before 6.4 and now want to use the new data loader? In other words, how can one use `unstable_HistoryRouter`  with `BrowserRouter`?

**Comment by sergioavazquez** [comment]
> @sergioavazquez Controlling navigation from outside the router is still possible, you just need to maintain a separate `history` dependency yourself (it's not a big problem). See: [#8264 (comment)](https://github.com/remix-run/react-router/issues/8264#issuecomment-991271554)

Why is it `unstable`? Are you going to remove this in a few months? I'm on the fence about using unstable methods on production. Also having to keep versions in check, so there's no incompatibility sounds like another point of failure. 

I'm sure Remix team removed it for a reason, why drop a working feature or label it unstable unless you foresee issues in the future? Also there's @9jaGuy 's question about using this with data loader.

This is a very common use case outside happy path documentation, it'd be great to get a fully supported method.

**Comment by Shaker-Pelcro** [comment]
There's the newly introduced "redirect" method to redirect from outside of the component.
I think it might help ...

https://reactrouter.com/en/main/fetch/redirect

**Comment by andrey779evseev** [comment]
> There's the newly introduced "redirect" method to redirect from outside of the component. it requires version 6.4 I think it might help ...
> 
> https://reactrouter.com/en/main/fetch/redirect

Sadly, it doesn't work this way `redirect('somewhere')` just return a Response object, and then loader does all work. But in my case I need to redirect user to login page from axios interceptors on 401 response status code. 

Something like this:
```
instance.interceptors.response.use(res => {
  return res
}, err => {
  if(err.response.status === 401) {
    // do some stuff for logout
    navigate('/login')
  }
  return Promise.reject(err)
})
```

**Comment by dathmart** [comment]
Meanwhile, in Remixland...
![image](https://user-images.githubusercontent.com/25598116/196052711-2fd07d47-1f3f-493c-b907-0411c225a689.png)

**Comment by Jamison1** [comment]
Aaaand It's Gone!

**unstable_HistoryRouter**
![Screen Shot 2022-10-16 at 3 29 01 PM](https://user-images.githubusercontent.com/7190240/196056817-b8a4c260-6841-4161-a509-41a97ecb5eb8.png)

[https://reactrouter.com/en/6.4.0-pre.14/routers/history-router](url)

Spent the past several days attempting to get 6.4 to work on a module federation mfe app without unstable_HistoryRouter (or unsafe_navigationcontext or using any imports from history) so I can use the new data loaders.

Trying with createBrowserRouter / createMemoryRouter for dev and prod.

Wondering if it's possible.

Some examples and up to date docs for 6.4.2 would be nice at this point.

**Comment by evoyy** [comment]
You can still use a custom history without even needing `unstable_HistoryRouter`.  If you dig into the source, you will see that a single global history instance is bound to the global window object. If you create a history outside of React:

````javascript
import { createBrowserHistory } from 'history';

let history = createBrowserHistory();
````

before rendering `<BrowserRouter>`, react router will use the history you have previously created yourself.

Obviously this is not supported, but it works.

**Comment by Jamison1** [comment]
> You can still use a custom history without even needing `unstable_HistoryRouter`. If you dig into the source, you will see that a single global history instance is bound to the global window object. If you create a history outside of React:
> 
> ```js
> import { createBrowserHistory } from 'history';
> 
> let history = createBrowserHistory();
> ```
> 
> before rendering `<BrowserRouter>`, react router will use the history you have previously created yourself.
> 
> Obviously this is not supported, but it works.

Yeah, this doesn't appear to be supported in the RR roadmap going forward (no imports from 'history' since 6.0-beta) and based on what I've tried to date it hasn't worked with the new data loaders.

In looking through all of the open and closed issues - this is one of the most active issues for RR right now that needs a solution inline with the roadmap.

**Comment by sergioavazquez** [comment]
> Aaaand It's Gone!
> 
> **unstable_HistoryRouter** ![Screen Shot 2022-10-16 at 3 29 01 PM](https://user-images.githubusercontent.com/7190240/196056817-b8a4c260-6841-4161-a509-41a97ecb5eb8.png)
> 
> https://reactrouter.com/en/6.4.0-pre.14/routers/history-router
> 

I knew it...

Whether it's a timeout or handling errors, we need a way to navigate outside components. I've been using this feature in all large apps I've worked in for at least 5 years now (some my design, some not). Not everything is a React component, What about Redux or Sagas? 

I'm happy to remove history as you say in the docs, but we need a way to navigate from outside components. This is not a minor detail I can just remove and adapt.

**Comment by candy4290** [comment]
How to navigate out component when use createBrowserRouter？
In this case, i can't use "<HistoryRouter history={history} basename='/'>"
help!!!!!!

**Comment by nigel-loops** [comment]
I am using router5 and was considering migrating my application to v6. The inability to navigate outside a component seems to disregard an enormous group of use cases and people fundamentally. I use redux-saga to manage complex flows. I know over the years, we have had other solutions to handle side effects, but with such an extensive application with 100+ saga files, this just isn't going to happen anytime soon.

At least router5 had a way to plug into the redux structure and allow navigations to happen outside of react components.

**Comment by Jamison1** [comment]
Found[ this](https://github.com/remix-run/react-router/issues/9351) explanation about the change in accessing history in another issue about RR upgrade from v6.3 to v6.4.  Still working through this issue!

![Yep! In react-router@6 3 (using history@5 ), there was no router per-se, just a navigator that triggered](https://user-images.githubusercontent.com/7190240/196964029-a11cd49e-bba9-451d-976d-670cf03cd84f.png)

**Comment by brophdawg11** [comment]
Hey folks!  I'm centralizing the info around `unstable_HistoryRouter` over in https://github.com/remix-run/react-router/issues/9422#issuecomment-1301182219 if you want to head over there for an explanation of where things stand!

**Comment by Vlad-Shevliakov** [comment]
There is no official solve of this issue still, why?

**Comment by ryanflorence** [comment]
React itself is the reason we discouraged navigating outside of the React tree due to UI sharding (a problem with React's concurrent mode and suspense where it would get out of sync with an external store like `history`).

To unblock our work in React Router, and to make sure we were forward compatible with React, we discouraged this use case and waited for the React team to provide a solution. However, we still provided a path forward for you with `unstable_HistoryRouter`.

It wasn't until React 18 that `useExternalSyncStore` was introduced, which is the solution we needed. It's what we've implemented in 6.4 with the new routers that are data aware.

We have a long history of maintaining our old versions. After 8+ years we've only had two breaking releases. Even v3 still gets patches. You can stay on v5 if things in v6 aren't worth the development effort to upgrade. It will be maintained.

If the features in v6 look interesting to you, the jump from an external `history` object and `unstable_HistoryRouter` to `createBrowserRouter` with the new `<RouterProvider>` should be trivial and you can use the new external `router` object to navigate outside of the react tree.

For more info, see the comment Matt linked to earlier.

**Comment by ryanflorence** [comment]
Conversation is locked so that further questions happen on the other issue :)

https://github.com/remix-run/react-router/issues/9422#issuecomment-1301182219

**Comment by ryanflorence** [comment]
Also, if you're on `unstable_HistoryRouter`, this should be a fairly straightforward change to your code to navigate outside of the react tree:

https://github.com/remix-run/react-router/issues/9422#issuecomment-1302564759
